content/2_Bronze/Intro_Sets.mdx

```diff
@@ -881,7 +881,7 @@ make their implementation easier.
 
   </Quiz.Question>
   <Quiz.Question>
-    What is the time complexity of insertions, deletions, and searches in an unordered map of size $N$?
+    What's the average time complexity of insertions, deletions, and searches in an unordered map of size $N$?
```

Comment:
averaged over what?

Reply:
idk, averaged over a ton of operations?
i feel like it should be clear from context

Reply:
You should make it clear what randomness you are averaging over. Either:
 - The sequence of operations is fixed, and the hash function is selected uniformly at random 
 - The hash function is fixed but the sequence of operations is randomized.

Reply:
this is in bronze
i don't think people will think that deeply about the complexity
it's mostly just "yeah this provides access fast enough"

Reply:
but it doesn't

Reply:
i mean sure if you go up high enough hashing might become a problem
but in bronze/silver?

Reply:
use `map` instead of `unordered_map` if you care about the time complexity

Reply:
what does that have to do with this question?
i mean i could just add a "given a uniformly random hash function?"

Reply:
> i mean i could just add a "given a uniformly random hash function"?

but `unordered_map` doesn't use a uniformly random hash function

> what does that have to do with this question?

If you don't think bronze/silver competitors should care about the time complexity of unordered map, then don't include this as a quiz question.

Reply:
> but `unordered_map` doesn't use a uniformly random hash function

what should the phrasing be then

> If you don't think bronze/silver competitors should care about the time complexity of unordered map, then don't include this as a quiz question.

99% of the time in cp problems an unordered map is O(1) with a pretty big constant factor. even though there are situations where it isn't, those should be covered in more advanced modules (https://usaco.guide/gold/hashmaps)

Reply:
> what should the phrasing be then

Either 
1. add "a uniformly random sequence of " before insertions
2. ask about worst-case instead of average case complexity

> 99% of the time in cp problems an unordered map is O(1) with a pretty big constant factor. 

I'm not sure what you mean by 99% of the time. On CF I doubt it's anywhere close to that.

Reply:
changed it, but what are these cf problems that don't allow unordered maps?

Reply:
not sure about normal CF rounds, but on educational CF these would be hacked all the time

Reply:
ok given this is the _usaco_ guide, i don't think hacks should count
i'm not doubting that given an adversary unordered maps would be constant time, but many people just get ac on the builtin test cases and call it a day

Reply:
> but many people just get ac on the builtin test cases and call it a day

If you try submitting `unordered_set` to the problem in the module, you'll get TLE.

Reply:
fair enough